contracts/VetoERC20Voting.sol

```diff
@@ -54,16 +77,107 @@ contract VetoERC20Voting is IVetoERC20Voting, TransactionHasher, Initializable {
+    function castFreezeVote() public {
+        uint256 userVotes;
+
+        if (
+            block.number >
+            freezeProposalCreatedBlock + freezeProposalBlockDuration
+        ) {
+            // Create freeze proposal, set total votes to msg.sender's vote count
+            freezeProposalCreatedBlock = block.number;
+
+            userVotes = votesToken.getPastVotes(
+                msg.sender,
+                freezeProposalCreatedBlock - 1
+            );
+            require(userVotes > 0, "User has no votes");
+
+            freezeProposalVoteCount = userVotes;
+
+            emit FreezeProposalCreated(msg.sender);
+        } else {
+            // There is an existing freeze proposal, count user's votes
+            require(
+                !userHasFreezeVoted[msg.sender][freezeProposalCreatedBlock],
+                "User has already voted"
+            );
+
+            userVotes = votesToken.getPastVotes(
+                msg.sender,
+                freezeProposalCreatedBlock - 1
+            );
+            require(userVotes > 0, "User has no votes");
+
+            freezeProposalVoteCount += userVotes;
+        }
+
+        userHasFreezeVoted[msg.sender][freezeProposalCreatedBlock] = true;
+
+        emit FreezeVoteCast(msg.sender, userVotes);
+    }
+
+    /// @notice Unfreezes the DAO, only callable by the owner
+    function defrost() public onlyOwner {
+        require(isFrozen(), "DAO is not already frozen");
+        freezeProposalCreatedBlock = 0;
+        freezeProposalVoteCount = 0;
+    }
+
+    /// @notice Updates the veto votes threshold, only callable by the owner
+    /// @param _vetoVotesThreshold The number of votes required to veto a transaction
+    function updateVetoVotesThreshold(uint256 _vetoVotesThreshold)
+        external
+        onlyOwner
+    {
+        vetoVotesThreshold = _vetoVotesThreshold;
     }
 
-    /// @notice Returns whether the specified functions has been vetoed
+    /// @notice Updates the freeze votes threshold, only callable by the owner
+    /// @param _freezeVotesThreshold Number of freeze votes required to activate a freeze
+    function updateFreezeVotesThreshold(uint256 _freezeVotesThreshold)
+        external
+        onlyOwner
+    {
+        freezeVotesThreshold = _freezeVotesThreshold;
+    }
+
+    /// @notice Updates the freeze proposal blocks duration, only callable by the owner
+    /// @param _freezeProposalBlockDuration The number of blocks a freeze proposal has to succeed
+    function updateFreezeProposalBlockDuration(
+        uint256 _freezeProposalBlockDuration
+    ) external onlyOwner {
+        freezeProposalBlockDuration = _freezeProposalBlockDuration;
+    }
+
+    /// @notice Updates the freeze block duration, only callable by the owner
+    /// @param _freezeBlockDuration The number of blocks a freeze last, from time of freeze proposal creation
+    function updateFreezeBlockDuration(uint256 _freezeBlockDuration)
```

Comment:
it seems like having this and `updateFreezeVotesThreshold` makes voting kind of strange, since users don't really know the parameters under which they're voting on a freeze.  If these are updated it should really start a new vote, I think

Reply:
This is a great point - If a parent DAO wants to update this state during the current voting period then it will change the current voting parameters.

This may be a better way to do it - however, this is how openzeppelin handles config updates performed by the DAO. Otherwise, we will have to keep a mapping of proposals w/ config details.

https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/governance/extensions/GovernorSettings.sol

Reply:
Hmm personally I don't see this as an issue. To change one of these parameters would require a proposal passed by the parent DAO, which are the same voters voting to freeze the DAO. 

I think the parent DAO _should_ have the ability to extend the freeze duration, even if a freeze is currently active.

Reply:
yeah, fair enough. I suppose to get around it they can just start another freeze vote.  The parent DAO really just has a whole lot of power over everything the child can do, which is interesting..